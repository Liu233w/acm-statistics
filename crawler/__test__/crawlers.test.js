const _ = require('lodash')

const poj = require('../crawlers/poj')
const hdu = require('../crawlers/hdu')
const zoj = require('../crawlers/zoj')
const dashiye = require('../crawlers/dashiye')
const codeforces = require('../crawlers/codeforces')
const uva = require('../crawlers/uva')
const fzu = require('../crawlers/fzu')
const spoj = require('../crawlers/spoj')
const timus = require('../crawlers/timus')
const leetcode_cn = require('../crawlers/leetcode_cn')
const vjudge = require('../crawlers/vjudge')
const csu = require('../crawlers/csu')
const {readConfigs} = require('../lib/configReader')

jest.setTimeout(10000) // 最多10秒

const notExistUsername = 'fmV84ZCQ3hwu'
const username = 'vjudge5'

// 另外，感谢 @leoloveacm, @2013300262 同学提供测试帐号

describe('poj', () => {

  test('test poj - 用户不存在时抛出异常', async () => {
    await expect(poj(null, notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test poj - 能够正确识别带有空格的用户名', async () => {
    await expect(poj(null, ' ' + notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test poj', async () => {
    const res = await poj(null, username)
    checkRes(res)
    expect(res.solvedList).toContain('1001')
  })

})

describe('hdu', () => {

  const config = {
    env: 'server',
  }

  test('test hdu - 用户不存在时抛出异常', async () => {
    await expect(hdu(config, notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test hdu - 能够正确识别带有空格的用户名', async () => {
    await expect(hdu(config, ' ' + notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test hdu', async () => {
    const res = await hdu(config, 'vjudge4')
    checkRes(res)
    expect(res.solvedList).toContain('1001')
  })

})

describe('zoj', () => {

  test('test zoj - 用户不存在时抛出异常', async () => {
    await expect(zoj(null, notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test zoj - 能够正确识别带有空格的用户名', async () => {
    await expect(zoj(null, ' ' + notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test zoj', async () => {
    const res = await zoj(null, username)
    checkRes(res)
    expect(res.solvedList).toContain('1001')
  })

})

describe('dashiye', () => {

  test('test dashiye - 用户不存在时抛出异常', async () => {
    await expect(dashiye(null, notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test dashiye - 能够正确识别带有空格的用户名', async () => {
    await expect(dashiye(null, ' ' + notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test dashiye', async () => {
    const res = await dashiye(null, username)
    checkRes(res)
    expect(res.solvedList).toContain('1001')
  })

})

describe('codeforces', () => {

  test('test codeforces - 用户不存在时抛出异常', async () => {
    await expect(codeforces(null, notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test codeforces - 能够正确识别带有空格的用户名', async () => {
    await expect(codeforces(null, ' ' + notExistUsername)).rejects.toThrow('handle: Field should contain only Latin letters, digits, underscore or dash characters')
  })

  test('test codeforces', async () => {
    const res = await codeforces(null, 'leoloveacm') // 没有找到好的测试多页返回的帐号，还是用这个测试单页吧
    checkRes(res)

    expect(new Set(res.solvedList)).toMatchObject(new Set([
      '754-B', '165-E', '492-E', '338-D', '333-A', '703-A', '540-C', '680-B', '680-A', '676-A', '450-B', '667-A', '667-C',
    ]))
  })

})

describe('uva', () => {

  test('test uva - 用户不存在时抛出异常', async () => {
    await expect(uva(null, notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test uva', async () => {
    const res = await uva(null, 'leoloveacm')
    checkRes(res)
    expect(res.solvedList).toContain('4141')
  })

})

describe('fzu', () => {

  test('test fzu - 用户不存在时抛出异常', async () => {
    await expect(fzu(null, notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test fzu', async () => {
    const res = await fzu(null, username)
    checkRes(res)
    expect(res.solvedList).toContain('1001')
  })

})

describe('spoj', () => {

  test('test spoj - 用户不存在时抛出异常', async () => {
    await expect(spoj(null, notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test spoj - 能够正确识别带有空格的用户名', async () => {
    await expect(spoj(null, ' ' + notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test spoj', async () => {
    const res = await spoj(null, username)
    checkRes(res)
    expect(res.solvedList).toContain('ABA12D')
  })

})

describe('timus', () => {

  test('test timus - 用户不存在时抛出异常', async () => {
    await expect(timus(null, notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test timus - 能够正确识别带有空格的用户名', async () => {
    await expect(timus(null, ' ' + notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test timus', async () => {
    const res = await timus(null, 'vjudge11')
    checkRes(res)
    expect(res.submissions).toBeGreaterThan(10)
    expect(res.solvedList).toContain('1387')
  })

})

describe('leetcode_cn', () => {

  test('test leetcode_cn - 用户不存在时抛出异常', async () => {
    await expect(leetcode_cn(null, notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test leetcode_cn - 能够正确识别带有空格的用户名', async () => {
    await expect(leetcode_cn(null, ' ' + notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test leetcode_cn', async () => {
    const res = await leetcode_cn(null, 'wwwlsmcom')
    checkRes(res)
  })

})

describe('csu', () => {

  test('test csu - 用户不存在时抛出异常', async () => {
    await expect(csu(null, notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test csu - 能够正确识别带有空格的用户名', async () => {
    await expect(csu(null, ' ' + notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test csu', async () => {
    const res = await csu(null, username)
    checkRes(res)
  })

})

describe('vjudge', () => {

  let vjConfig

  beforeAll(async () => {
    const config = await readConfigs()
    // 需要读取设置，因此没法在 ci 里面测试
    // 在进行测试之前首先要根据 README.md 里的说明来设置环境变量
    vjConfig = config.crawlers.vjudge
  })

  test('首先保证配置文件正确并且拥有正确的环境变量', () => {
    // 必须要保证配置文件正确才能进行此项测试
    expect(vjConfig).toBeTruthy()
    expect(vjConfig.crawler_login_user).toBeTruthy()
    expect(vjConfig.crawler_login_user).not.toBe('用户名')
    expect(vjConfig.crawler_login_password).toBeTruthy()
    expect(vjConfig.crawler_login_password).not.toBe('密码')

    // console.log(vjConfig)
  })

  test('test vjudge - 用户不存在时抛出异常', async () => {
    await expect(vjudge(vjConfig, notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test vjudge - 能够正确识别带有空格的用户名', async () => {
    await expect(vjudge(vjConfig, ' ' + notExistUsername)).rejects.toThrow('用户不存在')
  })

  test('test vjudge', async () => {
    const res = await vjudge(vjConfig, 'leoloveacm')
    checkRes(res)
    expect(res.solvedList).toContain('codeforces-436B')
  }, 50000)
})

function checkRes(res) {
  expect(typeof res.solved).toBe('number')
  expect(typeof res.submissions).toBe('number')

  expect(res.solved).toBeGreaterThan(0)
  expect(res.submissions).toBeGreaterThan(0)

  expect(res.submissions).toBeGreaterThan(res.solved)

  if (!_.isNil(res.solvedList)) {
    expect(res.solvedList.length).toBe(res.solved)
  }
}