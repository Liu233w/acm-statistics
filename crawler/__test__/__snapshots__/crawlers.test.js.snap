// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vjudge should work correctly 1`] = `
Object {
  "solved": 392,
  "solvedList": Array [
    "hdu-3966",
    "hdu-2063",
    "poj-1113",
    "poj-3264",
    "timus-1348",
    "dashiye-2243",
    "hdu-2243",
    "poj-1111",
    "poj-2236",
    "hdu-3473",
    "poj-2104",
    "hdu-4607",
    "hdu-3714",
    "hdu-1050",
    "hdu-1009",
    "hdu-2037",
    "hdu-1166",
    "hdu-2017",
    "LightOJ-1030",
    "hdu-1054",
    "poj-2311",
    "hdu-1907",
    "timus-1180",
    "LightOJ-1186",
    "hdu-1404",
    "LightOJ-1247",
    "LightOJ-1199",
    "LightOJ-1253",
    "LightOJ-1192",
    "hdu-4847",
    "poj-2318",
    "LightOJ-1004",
    "hdu-1848",
    "poj-3237",
    "spoj-QTREE",
    "dashiye-1036",
    "hdu-4825",
    "uvalive-7340",
    "codeforces-165E",
    "codeforces-492E",
    "codeforces-338D",
    "poj-1330",
    "poj-3468",
    "codeforces-327C",
    "codeforces-334C",
    "dashiye-1500",
    "dashiye-2002",
    "hdu-5058",
    "hdu-4456",
    "hdu-5692",
    "dashiye-1854",
    "poj-1151",
    "hdu-1394",
    "hdu-1556",
    "hdu-1541",
    "hdu-3038",
    "hdu-1506",
    "hdu-4496",
    "poj-1182",
    "hdu-5045",
    "hdu-3652",
    "hdu-2167",
    "poj-1185",
    "poj-2342",
    "hdu-4055",
    "hdu-3853",
    "codeforces-512B",
    "poj-2411",
    "hdu-2571",
    "hdu-2476",
    "hdu-2084",
    "hdu-1231",
    "OpenJ_Bailian-1088",
    "hdu-1494",
    "hdu-2191",
    "poj-1088",
    "hdu-5094",
    "hdu-2579",
    "codeforces-540C",
    "hdu-1495",
    "codeforces-429B",
    "hdu-4127",
    "hdu-1043",
    "poj-1077",
    "hdu-4403",
    "hdu-2510",
    "hdu-1175",
    "hdu-1455",
    "poj-1011",
    "codeforces-501E",
    "codeforces-442A",
    "hdu-2446",
    "codeforces-446B",
    "codeforces-490E",
    "codeforces-500C",
    "codeforces-478C",
    "codeforces-451B",
    "codeforces-462B",
    "codeforces-460B",
    "codeforces-456A",
    "poj-2387",
    "codeforces-442B",
    "codeforces-415A",
    "codeforces-369A",
    "poj-1651",
    "hdu-4763",
    "hdu-2896",
    "hdu-1251",
    "uva-11183",
    "hdu-4081",
    "hdu-2222",
    "hdu-3336",
    "hdu-1711",
    "hdu-4430",
    "poj-1961",
    "poj-3461",
    "poj-2752",
    "hdu-2112",
    "poj-2728",
    "poj-1287",
    "poj-2823",
    "hdu-3410",
    "hdu-5444",
    "poj-1577",
    "hdu-3415",
    "poj-2255",
    "codeforces-689D",
    "hdu-3791",
    "poj-1470",
    "hdu-5289",
    "uva-10090",
    "hdu-5443",
    "dashiye-1012",
    "hdu-2795",
    "poj-2182",
    "poj-2352",
    "poj-1064",
    "hdu-1538",
    "hdu-2177",
    "poj-1740",
    "hdu-2176",
    "hdu-1527",
    "hdu-1849",
    "hdu-1564",
    "poj-2484",
    "hdu-1847",
    "hdu-1846",
    "poj-3122",
    "codeforces-100989C",
    "codeforces-100989A",
    "poj-3301",
    "poj-3579",
    "poj-2456",
    "hdu-4349",
    "poj-1363",
    "hdu-3037",
    "poj-1845",
    "hdu-5673",
    "LightOJ-1370",
    "poj-2992",
    "hdu-4497",
    "hdu-1215",
    "LightOJ-1236",
    "poj-3090",
    "hdu-3001",
    "poj-2096",
    "poj-3254",
    "poj-1163",
    "poj-1157",
    "hdu-4389",
    "hdu-2089",
    "poj-2378",
    "poj-1463",
    "codeforces-472C",
    "codeforces-330C",
    "codeforces-330B",
    "codeforces-330A",
    "poj-2286",
    "fzu-2150",
    "codechef-DIGROT",
    "poj-1190",
    "uva-12558",
    "codechef-HOTEL",
    "poj-3126",
    "poj-1426",
    "poj-2251",
    "poj-1321",
    "poj-3278",
    "hdu-2897",
    "hdu-1730",
    "codeforces-388C",
    "codechef-DRAGNXOR",
    "hdu-2147",
    "hdu-3389",
    "poj-2368",
    "poj-2485",
    "poj-1861",
    "poj-1258",
    "poj-1177",
    "hdu-2193",
    "uvalive-4648",
    "hdu-4006",
    "poj-1251",
    "hdu-3068",
    "zoj-1610",
    "poj-2632",
    "poj-2528",
    "uvalive-6833",
    "uva-12718",
    "hdu-2064",
    "hdu-2013",
    "hdu-2010",
    "hdu-2032",
    "hdu-2012",
    "hdu-2026",
    "hdu-2020",
    "hdu-1164",
    "hdu-1754",
    "poj-1990",
    "poj-2299",
    "poj-2481",
    "hdu-4686",
    "codeforces-675C",
    "hdu-2629",
    "uva-10305",
    "poj-3067",
    "codeforces-450B",
    "hdu-1061",
    "LightOJ-1009",
    "uva-705",
    "uva-10050",
    "uva-540",
    "uva-11111",
    "poj-3070",
    "hdu-2035",
    "poj-1458",
    "LightOJ-1046",
    "LightOJ-1039",
    "hdu-4850",
    "codeforces-667C",
    "LightOJ-1012",
    "LightOJ-1037",
    "uva-297",
    "uva-784",
    "uva-572",
    "uva-657",
    "LightOJ-1032",
    "LightOJ-1033",
    "LightOJ-1038",
    "LightOJ-1044",
    "uva-401",
    "uva-10010",
    "uva-10361",
    "uva-548",
    "uva-442",
    "uva-673",
    "uva-10152",
    "uva-133",
    "uva-712",
    "uva-127",
    "uva-112",
    "uva-340",
    "uva-101",
    "LightOJ-1303",
    "LightOJ-1048",
    "LightOJ-1056",
    "hdu-1221",
    "zoj-2290",
    "timus-1792",
    "LightOJ-1212",
    "LightOJ-1113",
    "uva-10420",
    "LightOJ-1031",
    "LightOJ-1027",
    "LightOJ-1025",
    "LightOJ-1021",
    "LightOJ-1018",
    "LightOJ-1017",
    "LightOJ-1013",
    "LightOJ-1011",
    "poj-3125",
    "hdu-1021",
    "hdu-1285",
    "poj-2785",
    "poj-1979",
    "poj-1995",
    "poj-1006",
    "hdu-5612",
    "hdu-1045",
    "codeforces-651B",
    "codeforces-609A",
    "codeforces-615A",
    "codeforces-515C",
    "fzu-1327",
    "poj-2478",
    "codeforces-405B",
    "codeforces-580B",
    "hdu-4699",
    "poj-1013",
    "poj-1012",
    "hdu-4602",
    "hdu-4198",
    "hdu-3833",
    "hdu-3723",
    "hdu-4161",
    "hdu-4151",
    "hdu-4104",
    "hdu-3664",
    "poj-2362",
    "hdu-3579",
    "poj-2503",
    "zoj-2772",
    "zoj-2388",
    "hdu-2001",
    "hdu-2002",
    "hdu-2003",
    "hdu-2000",
    "poj-1915",
    "poj-1023",
    "poj-1056",
    "poj-3258",
    "poj-3273",
    "poj-3104",
    "hdu-1031",
    "hdu-1225",
    "hdu-4508",
    "hdu-4510",
    "hdu-4505",
    "poj-2243",
    "poj-2488",
    "hdu-1016",
    "poj-3190",
    "poj-3624",
    "poj-1061",
    "hdu-4864",
    "poj-3672",
    "poj-2272",
    "poj-3984",
    "poj-1562",
    "poj-1019",
    "poj-1007",
    "hdu-1896",
    "poj-2393",
    "poj-2782",
    "hdu-1023",
    "poj-1456",
    "hdu-2303",
    "poj-2833",
    "poj-2463",
    "hdu-1075",
    "hdu-1022",
    "hdu-2057",
    "hdu-4310",
    "hdu-1018",
    "hdu-1789",
    "hdu-1263",
    "hdu-2051",
    "hdu-2050",
    "hdu-1004",
    "hdu-1027",
    "hdu-1033",
    "poj-2242",
    "hdu-1228",
    "hdu-1029",
    "poj-1001",
    "poj-1573",
    "hdu-1035",
    "poj-3750",
    "uva-1225",
    "zoj-2482",
    "hdu-2098",
    "poj-2769",
    "poj-3006",
    "hdu-1019",
    "hdu-1030",
    "hdu-1034",
    "hdu-1013",
    "poj-2196",
    "hdu-1002",
    "hdu-2045",
    "codeforces-436B",
    "codeforces-387B",
    "codeforces-387A",
    "poj-2993",
    "poj-2996",
    "poj-1099",
    "hdu-2054",
    "hdu-1032",
    "codeforces-384A",
    "codeforces-399A",
    "SCU-4424",
    "hdu-2039",
  ],
  "submissions": 1282,
  "submissionsByCrawlerName": Object {
    "LightOJ": 71,
    "OpenJ_Bailian": 1,
    "SCU": 5,
    "codechef": 7,
    "codeforces": 109,
    "dashiye": 24,
    "fzu": 2,
    "hdu": 523,
    "poj": 398,
    "spoj": 1,
    "timus": 8,
    "uva": 111,
    "uvalive": 6,
    "zoj": 16,
  },
}
`;
